app/src/main/java/edu/usf/sas/pal/muser/util/EventCaptureUtils.java

```diff
@@ -0,0 +1,43 @@
+package edu.usf.sas.pal.muser.util;
+
+import android.content.Context;
+
+import com.simplecity.amp_library.ShuttleApplication;
+import com.simplecity.amp_library.model.Song;
+
+import edu.usf.sas.pal.muser.firebase.model.Event;
+import edu.usf.sas.pal.muser.firebase.model.SongData;
+
+
+
+public class EventCaptureUtils {
+
+    /**
+     * Function to populate the Event data class.
+     * @param song - The song on which the event happened.
+     * @param capturedEvent - The event that was captured.
+     * @param context - The context of the Fragment.
+     * @return - Event object
+     */
+    public static Event captureEvent(Song song, String capturedEvent, Context context){
+        long a = 0;
+        long b = 0;
+        SongData songData = populateSongData(song, context);
+        // TODO populate currentTimeMs, NanoTime, startTime and elaspedTime.
+        return new Event(capturedEvent, 0, 0, a,b, songData);
+    }
+
+    /**
+     * Function to populate SongData class. Called by captureEvent().
+     * @param song - Details of the song.
+     * @param context - Context of the fragment.
+     * @return - SongData object
+     */
+    public static SongData populateSongData(Song song, Context context){
+        return new SongData(song.id, song.name, song.artistId, song.albumName, song.albumId
```

Comment:
I'm still not convinced we're maximizing the null safety advantage of Kotlin here (i.e., we may be passing `null` values from `Song` to `SongData` for `?` fields), but let's go with this as-is and see what gets logged to Firebase for songs from different sources.

Reply:
Should I have a default value set in the data class for nullable fields so that they don't store `null` values?

Reply:
Yes, the nullable fields should have default values in case they aren't set via the constructor. But it's a little more complicated than that, especially when dealing with interop with Java. With the code as written, Java can still pass `null` values to the nullable fields, which would take the place of any default values.

In Kotlin you can have named parameters of a data class constructor that allow you to only populate some of the variables:

```kotlin
val jack = User(name = "Jack", age = 1)
```

...and the rest of the variables (e.g., `height`) would keep their default values.

Ideally we would know which of the `Song` fields could be null and which will always be populated. And for all fields that will always be populated, they shouldn't be nullable, and we should have a constructor just for those fields and then test for null for the remaining nullable fields and only set them if they aren't null. This way we would be guaranteed that we'd never need to null check the `SongData` fields.

One thing that may help here is to move this code extracting `Song` fields into `SongData` into a constructor in the `SongData` class that takes in the `Song` class. You could then try removing the `?` from all fields and Kotlin will then perform some static analysis to see if those fields could be null, and if so will give you a build error.

Reply:
Thank you for the detailed explanation on how we can maximize the Kotlin null safety advantage. I'm fairly new to Kotlin but I tried writing the code to extract the `Song` fields in the `SongData` class as you suggested. Please correct if I'm wrong about the approach. 

     (val id: Long,
     val name: String,
     val artistID: Long,
     val artistName: String,
     val albumID: Long,
     val albumName: String,
     val playlistID: Long,
     val playlistPlayOrder: Long?,
     val lastPlayed: Long?,
     val track: Int?,
     .............
    ){
           constructor(song: Song, context: Context): this(song.id, song.name, song.artistId, song.albumName, song.albumId,
            song.albumName, song.playlistSongId, song.playlistSongPlayOrder, song.lastPlayed,
            song.track, song.discNumber, song.getBitrateLabel(context), song.getFileSizeLabel(),
            song.isPodcast, song.duration, song.dateAdded, song.year, song.path, song.playCount,
            song.getSampleRateLabel(context), song.getFormatLabel())
      }


And I simply call the constructor:-  `SongData songData = new SongData(song, context)`

Reply:
Yes, the above looks right to me. I'd try removing the `?` for the variables and see if the compiler complains.

Reply:
I tried compiling the above code and the compiler didn't throw any warning or error. Does this mean that all the fields are always populated?

Reply:
If the entire app were pure Kotlin, then yes. But I'm not entirely sure how some of the compiler checks regarding null safety work with Java and Kotlin interop. Let's try keeping it this way for now. But we definitely need to test how this works with a value that's empty. So we could try removing the artist or song name from a tag and seeing if MUSER crashes.

Reply:
Just to clarify my above comment (I've since edited) - let's proceed with removing the `?` and test.

Reply:
That makes sense. Just to clarify, with the `?` removed, do you want me to use the `newSongData` function to populate the `SongData` class or should I use the constructor?

Reply:
Please use the constructor - you can remove the `newSongData` method from the utility class.